homeassistant/components/dsmr/sensor.py

```diff
@@ -754,21 +734,21 @@ def update_data(self, telegram: Telegram | None) -> None:
         """Update data."""
         self.telegram = telegram
         if self.hass and (
-            telegram is None or self.entity_description.obis_reference in telegram
+            telegram is None
+            or hasattr(telegram, self.entity_description.obis_reference)
         ):
             self.async_write_ha_state()
 
     def get_dsmr_object_attr(self, attribute: str) -> str | None:
         """Read attribute from last received telegram for this DSMR object."""
         # Make sure telegram contains an object for this entities obis
-        if (
-            self.telegram is None
-            or self.entity_description.obis_reference not in self.telegram
+        if self.telegram is None or not hasattr(
+            self.telegram, self.entity_description.obis_reference
         ):
             return None
 
         # Get the attribute value if the object has it
-        dsmr_object = self.telegram[self.entity_description.obis_reference]
+        dsmr_object = getattr(self.telegram, self.entity_description.obis_reference)
```

Comment:
Same as the original attempt to upgrade this library, we should avoid the use of getattr

Reply:
Well we cant. You have to do telegram.NAME or getattr(). the stuff in the dict is not the names, it is regexpes and their use is deprecated. It internally does setattr(telgram, name).

Reply:
https://github.com/ndokter/dsmr_parser/blob/94189b6b63e12554c866bc2964ffb7d5c7e8b2c4/dsmr_parser/objects.py#L30

For mbus variant there is no dict at all:
https://github.com/ndokter/dsmr_parser/blob/94189b6b63e12554c866bc2964ffb7d5c7e8b2c4/dsmr_parser/objects.py#L347

Reply:
If we make sure that the obis reference isn't computed in the entity description but declared, we could move the lookup to a lambda in the entity description that just accesses the correct attribute on the telegram instance.

Reply:
We would still need to check availability of the obis name, which would be either a hasattr or try except.

I suppose we could make the obis name a key in a dict, and iterate over the names in the telegram.

But we dont gain much since there is no typing info for the attributes, they are dynamically added, and there are no constants defined for the names.

Do you still think that would be better?

Reply:
Actually would still need a hasattr or catch attribute exception on most telegrams since not all telegrams contain all attributes.

The behavior of the telegram object is still much more like a dict, but its using attributes instead of dict keys. So its very much akin to .get() on a dict.

Reply:
<img width="823" alt="image" src="https://github.com/user-attachments/assets/c7fd70c4-e2d3-427b-9034-2d9dde0061d7">

Since it's a typed library, and the attribute is dynamically added, we can't even access the attribute without a type warning.

Reply:
The design of the library seems flawed for type annotations.

With the current design of the library, I agree that it doesn't seem useful to use named attribute access.

Reply:
Meh, not sure why they deprecate the old for this method upstream in that case, as it is not really an improvement from the perspective of the consumer of this library.

Reply:
Nope, imho just worse. The old code was/is weird too though. It used regexp strings as keys in the dict, meaning they could not print the data with actual names.

I did add a test that ensure we dont reference any unknown obis names, so ive mitigated the risk of unexpected breakage on later updates(found one if my own in the process)

Reply:
Can we move forward with this or do you guys have done other ideas?

Reply:
No makes sense, there is simply no better way out ATM. 👍